src/v/cloud_storage/topic_manifest.cc

```diff
@@ -352,53 +375,87 @@ void topic_manifest::do_update(const topic_manifest_handler& handler) {
+            vlog(cst_log.debug, "Parsed successfully!");
+            topic_manifest::do_update(handler);
         } else {
-            throw std::runtime_error(fmt_with_ctx(
-              fmt::format,
-              "Failed to parse topic manifest: {} at offset {}",
-              rapidjson::GetParseError_En(e),
-              o));
+            rapidjson::ParseErrorCode e = reader.GetParseErrorCode();
+            size_t o = reader.GetErrorOffset();
+
+            if (_topic_config) {
+                throw std::runtime_error(fmt_with_ctx(
+                  fmt::format,
+                  "Failed to parse topic manifest {}: {} at offset {}",
+                  get_manifest_path(),
+                  rapidjson::GetParseError_En(e),
+                  o));
+            } else {
+                throw std::runtime_error(fmt_with_ctx(
+                  fmt::format,
+                  "Failed to parse topic manifest: {} at offset {}",
+                  rapidjson::GetParseError_En(e),
+                  o));
+            }
         }
+        break;
+    }
+    case manifest_format::serde:
+        // serde format is straightforward: the buffer is a
+        // topic_manifest_state
+        auto m_state = serde::from_iobuf<topic_manifest_state>(
+          std::move(result));
+        _topic_config = std::move(m_state.cfg);
+        _rev = m_state.initial_revision;
+        _manifest_version = topic_manifest::serde_version;
+        break;
     }
+
     co_return;
 }
 
 ss::future<serialized_data_stream> topic_manifest::serialize() const {
-    iobuf serialized;
-    iobuf_ostreambuf obuf(serialized);
-    std::ostream os(&obuf);
-    serialize(os);
-    if (!os.good()) {
-        throw std::runtime_error(fmt_with_ctx(
-          fmt::format,
-          "could not serialize topic manifest {}",
-          get_manifest_path()));
+    vassert(_topic_config.has_value(), "_topic_config is not initialized");
+
+    if (_manifest_version == first_version) {
+        // serialize in json format. this could still be required in unit-tests
+        // and mixed-versions clusters, when the old node is not yet running
+        // with features::feature::cluster_topic_manifest_format_v2
+        iobuf serialized;
+        iobuf_ostreambuf obuf(serialized);
+        std::ostream os(&obuf);
+        serialize(os);
+        if (!os.good()) {
+            throw std::runtime_error(fmt_with_ctx(
+              fmt::format,
+              "could not serialize topic manifest {}",
+              get_manifest_path()));
+        }
+        size_t size_bytes = serialized.size_bytes();
+        co_return serialized_data_stream{
+          .stream = make_iobuf_input_stream(std::move(serialized)),
+          .size_bytes = size_bytes};
     }
-    size_t size_bytes = serialized.size_bytes();
+
+    // serialize in binary format
+    auto serialized = serde::to_iobuf(topic_manifest_state{
+      .cfg = _topic_config.value(), .initial_revision = _rev});
```

Comment:
I don't imagine it'd be majorly problematic, but maybe it makes sense to have the state be a member of the manifest so we avoid this copy

Reply:
yeah it would not be wrong
the difference would be if we are ok with serialize() moving its internal state to to_iobuf. since this it not a && method, i went with copy

Reply:
Hm I don't understand: to_iobuf uses perfect forwarding -- isn't that sufficient to let us copy transparently?